File: find-minimum-diameter-after-merging-two-trees.cpp
class Solution {
public:
    int diameter(vector<vector<int>>& edges) {
        int n = edges.size() + 1;
        vector<set<int>> graph(n);
        for (const auto& edge : edges) {
            graph[edge[0]].insert(edge[1]);
            graph[edge[1]].insert(edge[0]);
        }
        vector<int> sizes(n, 0);
        auto dfs = [&](auto&& dfs, int i, int prev) -> int {
            sizes[i] = 1;
            for (int j : graph[i]) {
                if (j != prev) {
                    sizes[i] += dfs(dfs, j, i);
                }
            }
            return sizes[i];
        };
        auto getcentroid = [&](auto&& getcentroid, int i, int prev, int total) -> int {
            for (int j : graph[i]) {
                if (j != prev && sizes[j] > total / 2) {
                    return getcentroid(getcentroid, j, i, total);
                }
            }
            return i;
        };
        auto longest = [&](auto&& longest, int i, int prev) -> int {
            int res = 1;
            for (int j : graph[i]) {
                if (j != prev) {
                    res = max(res, 1 + longest(longest, j, i));
                }
            }
            return res;
        };
        auto solve = [&](auto&& solve, int i) -> int {
            int total = dfs(dfs, i, -1);
            int centroid = getcentroid(getcentroid, i, -1, total);
            int a = 0, b = 0;
            for (int j : graph[centroid]) {
                int longest_val = longest(longest, j, centroid);
                if (longest_val > a) {
                    b = a;
                    a = longest_val;
                } else if (longest_val > b) {
                    b = longest_val;
                }
            }
            int res = a + b;
            for (int j : graph[centroid]) {
                graph[j].erase(centroid);
                res = max(res, solve(solve, j));
            }
            return res;
        };
        return solve(solve, 0);
    }

    int minimumDiameterAfterMerge(vector<vector<int>>& edges1, vector<vector<int>>& edges2) {
        int a = diameter(edges1);
        int b = diameter(edges2);
        return max({a, b, (a + 1) / 2 + (b + 1) / 2 + 1});
    }
};

File: find-minimum-diameter-after-merging-two-trees.py
class DSU:
    def __init__(self, n):
        self.parent = list(range(n))
        self.diameters = [(i, i) for i in range(n)]
        self.size = [1] * n
        
    def find(self, a):
        if a == self.parent[a]:
            return a
        self.parent[a] = self.find(self.parent[a])
        return self.parent[a]
        
    def union(self, a, b):
        parent_a = self.find(a)
        parent_b = self.find(b)
        if parent_a != parent_b:
            if self.size[parent_a] < self.size[parent_b]:
                parent_a, parent_b = parent_b, parent_a
            self.parent[parent_b] = parent_a
            self.size[parent_a] += self.size[parent_b]

class Solution:
    def diameter(self, edges):
        n = len(edges) + 1
        graph = [[] for _ in range(n)]
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        depth = [0] * n
        LOG = 18
        p = [[i] * LOG for i in range(n)]
        stack = [(0, -1, 0)]
        while stack:
            i, pr, d = stack.pop()
            depth[i] = d
            for j in graph[i]:
                if j != pr:
                    p[j][0] = i
                    stack.append((j, i, d + 1)) 
        for l in range(1, LOG):
            for i in range(n):
                p[i][l] = p[p[i][l - 1]][l - 1]
        dsu = DSU(n)
        ldiameter = 0    
        def addEdge(u, v):
            nonlocal ldiameter
            l = max(dsu.diameters[dsu.find(u)], key = lambda x: getDist(u, x))
            r = max(dsu.diameters[dsu.find(v)], key = lambda x: getDist(v, x))
            first, last = max([(l, r), dsu.diameters[dsu.find(u)], dsu.diameters[dsu.find(v)]], key = lambda e: getDist(e[0], e[1]))
            dsu.union(u, v)
            dsu.diameters[dsu.find(u)] = (first, last)
            ldiameter = max(ldiameter, getDist(first, last))   
        def getDist(u, v):
            res = depth[u] + depth[v]
            if depth[u] > depth[v]:
                u, v = v, u
            diff = depth[v] - depth[u]
            jump = 0
            while diff:
                if diff & 1:
                    v = p[v][jump]
                jump += 1
                diff //= 2
            for jump in range(LOG - 1, -1, -1):
                if p[u][jump] != p[v][jump]:
                    u = p[u][jump]
                    v = p[v][jump]
            LCA = p[u][0] if u != v else u
            res -= 2 * depth[LCA]
            return res
        for u, v in edges:
            addEdge(u, v)
        return ldiameter
 
    def minimumDiameterAfterMerge(self, edges1: List[List[int]], edges2: List[List[int]]) -> int:
        a = self.diameter(edges1)
        b = self.diameter(edges2)
        return max(a, b, (a + 1) // 2 + (b + 1) // 2 + 1)